File: phyphox/ExperimentAudioInput.m
//
//  ExperimentAudioInput.m
//  phyphox
//
//  Created by Jonas Gessner on 23.03.16.
//  Copyright © 2016 RWTH Aachen. All rights reserved.
//

#import "ExperimentAudioInput.h"
#import "phyphox-swift.h"

@implementation ExperimentAudioInput {
    EZMicrophone *_microphone;
    BOOL _recording;
}

- (instancetype)initWithSampleRate:(NSUInteger)sampleRate outBuffers:(NSArray <DataBuffer *>*)outBuffers {
    self = [super init];
    
    if (self) {
        _sampleRate = sampleRate;
        _outBuffers = [outBuffers copy];
    }
    
    return self;
}

- (void)startRecording {
    if (_recording) {
        return;
    }
    
    _recording = YES;
    
    if (_microphone == nil) {
        _microphone = [[EZMicrophone alloc] initWithMicrophoneDelegate:self withAudioStreamBasicDescription:[EZAudioUtilities monoFloatFormatWithSampleRate:_sampleRate]];
    }
    
    [_microphone startFetchingAudio];
}

- (void)stopRecording {
    if (!_recording) {
        return;
    }
    
    [_microphone stopFetchingAudio];
    
    
    _recording = NO;
}

- (void)microphone:(EZMicrophone *)microphone hasAudioReceived:(float **)buffer withBufferSize:(UInt32)bufferSize withNumberOfChannels:(UInt32)numberOfChannels {
    float *values = buffer[0];
    
    NSMutableArray <NSNumber *>* final = [NSMutableArray arrayWithCapacity:bufferSize];
    
    for (int i = 0; i < bufferSize; i++) {
        float value = values[i];
        [final addObject:@(value)];
    }
    
    for (DataBuffer *outBuffer in _outBuffers) {
        [outBuffer appendFromArray:final notify:YES];
    }
}

@end

File: phyphox/ExperimentAudioInput.swift
//
//  ExperimentAudioInput.swift
//  phyphox
//
//  Created by Jonas Gessner on 13.01.16.
//  Copyright © 2016 RWTH Aachen. All rights reserved.
//

import Foundation

final class ExperimentAudioInput {
    let sampleRate: UInt
    let outBuffers: [DataBuffer]
    
    var receiver: AEBlockAudioReceiver!
    
    init(sampleRate: UInt, outBuffers: [DataBuffer]) {
        self.sampleRate = sampleRate
        self.outBuffers = outBuffers
        
        defer {
            receiver = AEBlockAudioReceiver { [unowned self] (from: UnsafeMutablePointer<Void>, timestamp: UnsafePointer<AudioTimeStamp>, frames: UInt32, data: UnsafeMutablePointer<AudioBufferList>) in
                autoreleasepool({
                    var array = [Float](count: Int(frames), repeatedValue: 0.0)
                    
                    var arrayPointer = UnsafeMutablePointer<Float>(array)
                    
                    defer {
                        arrayPointer.destroy()
                    }
                    
                    AEFloatConverterToFloat(ExperimentManager.sharedInstance().floatConverter, data, &arrayPointer, frames)
                    
                    let final = array.map{ Double($0) }
                    
                    dispatch_async(dispatch_get_main_queue()) {
                        for out in self.outBuffers {
                            out.appendFromArray(final)
                        }
                    }
                })
            }
        }
    }
    
    func startRecording(experiment: Experiment) {
        ExperimentManager.sharedInstance().audioController.addInputReceiver(receiver)
    }
    
    func stopRecording(experiment: Experiment) {
        ExperimentManager.sharedInstance().audioController.removeInputReceiver(receiver)
    }
}
